design/layer2-prefer-leader.md

```diff
@@ -0,0 +1,103 @@
+# Configuring Preferred Node To Be Layer2 Leader
+
+## Summary
+
+The purpose of the ehancement is to support configuring preferred node to be the leader for L2 mode.
+
+With this enhancement, the leader election will be artificially controllable.
+
+## Motivation
+
+MetalLB’s layer2 mode is simliar to `Keepalived`(https://metallb.universe.tf/concepts/layer2/#comparison-to-keepalived) which using `VRRP`.
+
+This enhancement implement functionality similar to [Priority and Preemption](https://www.cisco.com/assets/sol/sb/Switches_Emulators_v2_3_5_xx/help/350_550/index.html#page/tesla_350_550_olh/ts_vrrp_18_09.html) configuration of `VRRP`.
+
+### Goals
+
+- In L2 mode, being able to specify a node as the preferred leader for a service.
+- When a node is configured as the preferred leader, it replaces the current leader.
+
+### Non-Goals
+
+- Change the current leader mechanism election using memberlist.
+- Change the data models of Metallb's CRDs.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+When using Metallb's Layer2 mode, user wants to specify one node as the leader for a service and let all traffic for this service's IP goes to this sepcified node.
+
+#### Story 2
+
+When using Metallb's Layer2 mode, the current leader node requires maintenance, user want to manually switchover to a sepcified node.
+
+## Design Details
+
+### Preferred Node Design proposal
+
+The proposal here is to:
+
+1. Add preferred node configuration for a service. In order not to change exsiting data model of the CRDs, 
+   the configuration can be placed in an annotation of the service. This can also make the configuration 
+   service-oriented and more flexible.
+
+2. Enhance Metallb's layer2 leader election process, choose the preferred node as leader.
+
+#### Preferred Node configuration
+
+If user prefer one node to serve the Load Balancer IP, add the `metallb.universe.tf/preferredNode` annotation to service,
```

Comment:
We still need to settle the discussion about the rationale, but I'd rather have this part of the l2advertisement which is in control of the cluster admin, rather than being part of the service definition which is handled by the application.

Reply:
It is a matter of granularity. We consider giving user a choice to assign the preferred node is more feasible, so different users won't have to share the same configuration of l2advertisement.  User A may prefer to use node X as his preferred node for this service,  while user B may prefer to use node Y as his, and both services can be served by the same l2advertisement/IPAddressPool.

Reply:
Yes, but this choice can't be left to the user, because you'd risk to have all the services pinned to a node. This setting must be under control of the administrator.